counterparty-core/counterpartycore/lib/messages/move.py

```diff
@@ -0,0 +1,84 @@
+import logging
+
+from counterpartycore.lib import config, exceptions, ledger, script, util
+
+logger = logging.getLogger(config.LOGGER_NAME)
+
+
+def compose(db, source, destination):
+    # We don't check if the source has attached assets
+    # to allow transaction chaining in the same block
+
+    if not util.is_utxo_format(source):
+        raise exceptions.ComposeError("Invalid source utxo format")
+
+    try:
+        script.validate(destination)
+    except script.AddressError as e:
+        raise exceptions.ComposeError("destination must be an address") from e
+
+    return (source, [(destination, None)], None)
+
+
+# call on each transaction
+def move_assets(db, tx):
+    if "utxos_info" not in tx or not tx["utxos_info"]:
+        return
+
+    sources, destination, _outputs_count, _op_return_output = util.parse_utxos_info(
+        tx["utxos_info"]
+    )
+
+    # do nothing if no source or destination
+    if len(sources) == 0 or not destination:
+        return
+
+    # we move all assets from the sources to the destination
+    action = "utxo move"
+    msg_index = 0
+    # we move all assets from each source to the destination
+    for source in sources:
+        balances = ledger.get_utxo_balances(db, source)
+        for balance in balances:
+            if balance["quantity"] == 0:
+                continue
+            # debit asset from source
+            ledger.debit(
```

Comment:
In blocks.py you have attach.parse and detach.parse.
When looking in detach, parse goes over all sources and calls detach, and detach debits/credits the ledger.
then in blocks.py, you call move.move_assets which once again debits/credits the ledger.

It looks like the assets are being moved twice? What am I missing?

Reply:
you are absolutely right! see the comment here: https://github.com/CounterpartyXCP/counterparty-core/pull/2626/files#diff-d605bdc22d2945dfe8e33a259ec63f2bda3d3a9a9ea8aea1022ec17541224b3dR141

The assets will not be debited twice, nothing will happen at all because the balances will already be emptied by the detach.
On the other hand, I think it is better to leave the call of `utxo.move()` after the transaction parsing even for the detach because if the detach is invalid (for example with an invalid destination) there is no reason not to do a possible utxo move.

Reply:
What if there’s no NON-OPRETURN in an invalid destination detach transaction? In that case, it seems the assets would remain in spent UTXOs, which we should aim to avoid after this fix. I suggest detaching the assets to the origin input addresses, similar to how it’s handled when no destination is specified in `detach.py`.

Reply:
with the new `utxos_info` field format we don't check the outputs at all for detach. All the inputs are detached to the  provided destination or, if not provided, by default to the address corresponding to the vin.

Reply:
I'm not sure we should fallback to the address corresponding to the vin if the destination is invalid. I feel is cleaner to mark the transaction as invalid and do nothing. Could be a typo and the user will fix in a new tx for example.

Reply:
> I'm not sure we should fallback to the address corresponding to the vin if the destination is invalid. I feel is cleaner to mark the transaction as invalid and do nothing. Could be a typo and the user will fix in a new tx for example.

They can’t perform any action with a new transaction because the asset UTXO has already been spent.

Reply:
yeah make sense!